platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileReaderPbj.java

```diff
@@ -0,0 +1,357 @@
+                buf = readingMmaps.get(index);
+            }
+        }
+        BufferedData readBuf = readingBufs.get(index);
+        if ((readBuf == null) || (readBuf.capacity() != buf.capacity())) {
+            final BufferedData newReadBuf = BufferedData.wrap(buf);
+            if (readingBufs.compareAndSet(index, readBuf, newReadBuf)) {
+                readBuf = newReadBuf;
+            } else {
+                readBuf = readingBufs.get(index);
+            }
+        }
+        return readBuf;
+    }
+
+    private void openNewFileChannel() throws IOException {
+        assert (fileChannel == null) || !fileChannel.isOpen();
+        fileChannel = FileChannel.open(path, StandardOpenOption.READ);
+    }
+
+    /**
+     * Read bytesToRead bytes of data from the file starting at byteOffsetInFile unless we reach the
+     * end of file. If we reach the end of file then returned buffer's limit will be set to the
+     * number of bytes read and be less than bytesToRead.
+     *
+     * @param byteOffsetInFile Offset to start reading at
+     * @return ByteBuffer containing read data. This is a reused per thread buffer, so you can use
+     *     it till your thread calls read again.
+     * @throws IOException if there was a problem reading
+     * @throws ClosedChannelException if the file was closed
+     */
+    private BufferedData read(final long byteOffsetInFile) throws IOException {
+        final int bufIndex = Math.toIntExact(byteOffsetInFile / (MMAP_BUF_SIZE / 2));
+        // Try a few times. It's very unlikely (other than in tests) that a thread is
+        // interrupted more than once in short period of time, so 2 retries should be enough
+        final long mmapOffset = (long) bufIndex * (MMAP_BUF_SIZE / 2);
+        for (int retries = 2; retries > 0; retries--) {
+            try {
+                final BufferedData in = getReadBuffer(bufIndex);
+                final int tag = in.getVarInt(byteOffsetInFile - mmapOffset, false); // tag
+                assert tag == ((FIELD_DATAFILE_ITEMS.number() << TAG_FIELD_OFFSET) | WIRE_TYPE_DELIMITED);
+                final int sizeOfTag = ProtoUtils.sizeOfUnsignedVarInt32(tag);
+                final int size = in.getVarInt(byteOffsetInFile + sizeOfTag - mmapOffset, false);
+                if (size >= MMAP_BUF_SIZE / 2) {
+                    throw new UnsupportedOperationException("Data item is too large");
+                }
+                final int sizeOfSize = ProtoUtils.sizeOfUnsignedVarInt32(size);
+                BufferedData buf = BUFFER_CACHE.get();
+                if ((buf == null) || (buf.capacity() < size)) {
+                    buf = BufferedData.allocate(size);
+                    BUFFER_CACHE.set(buf);
+                }
+                buf.position(0);
+                buf.limit(size);
+                in.getBytes(byteOffsetInFile + sizeOfTag + sizeOfSize - mmapOffset, buf);
+//                buf.flip(); // TODO: why isn't buf position updated?
```

Comment:
Don't forget to clean it up.

Reply:
There is a PBJ ticket about it, https://github.com/hashgraph/pbj/issues/78. I added the link here